datalad/tests/test_annexrepo.py

```diff
@@ -0,0 +1,36 @@
+# emacs: -*- mode: python-mode; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Tests to test implementation of class AnnexRepo
+
+Note: There's not a lot to test by now.
+
+"""
+__author__ = 'Benjamin Poldrack'
+
+import os.path
+from shutil import rmtree
+
+from nose.tools import assert_raises, assert_is_instance, assert_true
+from git.exc import GitCommandError
+
+from datalad.support.annexrepo import AnnexRepo
+from datalad.tests.utils import with_tempfile, with_testrepos, assert_cwd_unchanged
+
+
+@assert_cwd_unchanged
+@with_testrepos(flavors=['local'])
+@with_tempfile
+def test_AnnexRepo(src, dst):
+
+    ar = AnnexRepo(dst, src)
+    assert_is_instance(ar, AnnexRepo, "AnnexRepo was not created.")
+    assert_true(os.path.exists(os.path.join(dst, '.git', 'annex')))
+
+    #do it again should raise GitCommandError since git will notice there's already a git-repo at that path
+    assert_raises(GitCommandError, AnnexRepo, dst, src)
```

Comment:
>You are welcome to add default "src=None" argument to __init__ and if not
none -- run "clone".

That's the way it already is?!

>AnnexRepo class probably should just take "dst" as the argument, and then
have a method "clone" which would get it actually installed if not there
yet or puke if there already

What is the advantage? Basically, this is what happens. Only difference is,
that you want the extra clone-method. But: What for? When do you want to
clone if not on instantiating an object of this class?
I don't see, when it makes sense to call that clone-method from outside of
the constructor.


Regarding several instances representing the same repo - not sure whether
or not this may lead to problems later on. By now, you are right. No
problem, but also no use for, as far as I can see.

2015-02-25 15:00 GMT+01:00 Yaroslav Halchenko <notifications@github.com>:

> In datalad/tests/test_annexrepo.py
> <https://github.com/datalad/datalad/pull/42#discussion_r25340967>:
>
> > +
> > +from datalad.support.annexrepo import AnnexRepo
> > +from datalad.tests.utils import with_tempfile, with_testrepos, assert_cwd_unchanged
> > +
> > +
> > +@assert_cwd_unchanged
> > +@with_testrepos(flavors=['local'])
> > +@with_tempfile
> > +def test_AnnexRepo(src, dst):
> > +
> > +    ar = AnnexRepo(dst, src)
> > +    assert_is_instance(ar, AnnexRepo, "AnnexRepo was not created.")
> > +    assert_true(os.path.exists(os.path.join(dst, '.git', 'annex')))
> > +
> > +    #do it again should raise GitCommandError since git will notice there's already a git-repo at that path
> > +    assert_raises(GitCommandError, AnnexRepo, dst, src)
>
> ok,.... I think logic is not right here.... AnnexRepo class probably
> should just take "dst" as the argument, and then have a method "clone"
> which would get it actually installed if not there yet or puke if there
> already. If annex repo exists at 'dst' it should be ok to create AnnexRepo
> instance again. the same goes for the Dataset.
> You are welcome to add default "src=None" argument to __init__ and if not
> none -- run "clone".
>
> Alternative to the former, we might use chaining pattern. Michael opposed
> something like that in PyMVPA but I think it is quite neat and other
> toolkits use it as well
>
> class C(object):
>     def __init__(self, d):
>         self.d = d
>
>     def clone(self, src):
>         print("cloning %s into %s" % (src, self.d))
>         return self
>
>     def get(self, f):
>         print("Getting file %s" % f)
>         return self
>
> C('/tmp/test').clone('origin').get('data.dat')
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/datalad/datalad/pull/42/files#r25340967>.
>